lahja/base.py

```diff
@@ -157,38 +158,122 @@ class BaseRemoteEndpoint(RemoteEndpointAPI):
+        return f"RemoteEndpoint[{self._name if self._name is not None else id(self)}]"
 
     def __repr__(self) -> str:
         return f"<{self}>"
 
+    _name: Optional[str] = None
+
+    @property
+    def name(self) -> str:
+        if self._name is None:
+            raise AttributeError("Endpoint name has not yet been established")
+        else:
+            return self._name
+
     async def wait_started(self) -> None:
         await self._running.wait()
 
+    async def wait_ready(self) -> None:
+        await self._ready.wait()
+
     async def wait_stopped(self) -> None:
         await self._stopped.wait()
 
     def is_running(self) -> bool:
         return not self.is_stopped and self.running.is_set()
 
+    def is_ready(self) -> bool:
+        return self.is_running and self._ready.is_set()
+
     def is_stopped(self) -> bool:
         return self._stopped.is_set()
 
+    async def _run(self) -> None:
+        self._running.set()
+
+        # Send the hello message
+        await self.send_message(Hello(self._local_name))
+
+        # Wait for the other endpoint to identify itself.
+        hello = await self.conn.read_message()
+        if isinstance(hello, Hello):
+            self._name = hello.name
+            self.logger.debug(
+                "RemoteEndpoint connection established: %s <-> %s",
+                self._local_name,
+                self.name,
+            )
+            self._ready.set()
+        else:
+            self.logger.debug(
+                "Invalid message: First message must be `Hello`, Got: %s", hello
+            )
+            self._stopped.set()
+            return
+
+        while self.is_running:
+            try:
+                message = await self.conn.read_message()
+            except RemoteDisconnected:
+                async with self._received_response:
+                    self._received_response.notify_all()
+                return
+
+            if isinstance(message, Broadcast):
+                await self.new_msg_func(message)
+            elif isinstance(message, SubscriptionsUpdated):
+                self._subscriptions_initialized.set()
+                async with self._received_subscription:
+                    self._subscribed_events = message.subscriptions
+                    self._received_subscription.notify_all()
+                # The ack is sent after releasing the lock since we've already
+                # exited the code which actually updates the subscriptions and
+                # we are merely responding to the sender to acknowledge
+                # receipt.
+                if message.response_expected:
+                    await self.send_message(SubscriptionsAck())
+            elif isinstance(message, SubscriptionsAck):
+                async with self._received_response:
+                    self._received_response.notify_all()
+            else:
+                self.logger.error(f"received unexpected message: {message}")
```

Comment:
The body of this `while` loop is unchanged, only moved from it's previous location in `lahja.endpoint.asyncio`